File: depgraph/extractors/generic/python/detectors/service.py
"""Detect service-layer functions.

A function is labeled kind=service if it lives under a directory named
`services` or `utils` (case-sensitive, at any depth) and is:
- module-level (no enclosing class),
- public (does not start with `_`),
- not already labeled by another detector running after this one.

This is a path-convention detector, not framework-detected. Lifted
from Concorda's extract_api.py extract_service_nodes().
"""
from __future__ import annotations

import ast
from typing import Any
from pathlib import PurePosixPath

from extractors.generic.python.detector_api import (
    Detector, DetectorContext, Mutation, RelabelNode,
)


_SERVICE_DIRS = {"services", "utils"}


def _is_service_path(path: str) -> bool:
    parts = PurePosixPath(path).parts
    return any(p in _SERVICE_DIRS for p in parts)


class ServiceDetector(Detector):
    name = "service"

    def detect(
        self,
        tree: ast.AST,
        primitives: list[dict[str, Any]],
        ctx: DetectorContext,
    ) -> list[Mutation]:
        if not _is_service_path(ctx.file_path):
            return []
        # Pre-flip iterates tree.body (top-level only); match that.
        if not isinstance(tree, ast.Module):
            return []
        # Map top-level function name -> args list from the AST so we
        # emit args metadata that mirrors extract_api.py:710.
        args_by_name: dict[str, list[str]] = {}
        for node in tree.body:
            if isinstance(node, (ast.FunctionDef, ast.AsyncFunctionDef)):
                args_by_name[node.name] = [a.arg for a in node.args.args]

        muts: list[Mutation] = []
        for n in primitives:
            if n["kind"] != "function":
                continue
            if n.get("parent_id") is not None:
                # method, skip
                continue
            if n["name"].startswith("_"):
                continue
            if n["name"] not in args_by_name:
                # Not a true top-level def (e.g. nested-but-emitted-at-module).
                continue
            muts.append(RelabelNode(
                node_id=n["id"],
                new_kind="service",
                metadata={"args": args_by_name[n["name"]]},
            ))
        return muts

File: depgraph/extractors/generic/typescript/detectors/service.ts
/**
 * Service detector — emits service AddNode candidates for top-level
 * functions and arrow consts that live under lib/, pages/, services/, or
 * utils/ (at any depth). Skips private (`_`-prefixed) names.
 *
 * Emits the same _-prefixed canonical metadata as react.ts so the shared
 * canonicalize() stage in extract.ts can build the canonical service node.
 *
 * Path convention only — no framework recognition. Lifted from Concorda's TS
 * extractor.
 */
import * as ts from "typescript";
import { Node } from "ts-morph";
import {
  Detector, DetectorContext, Mutation, Primitive,
} from "../detector_api.js";

const SERVICE_DIRS = new Set(["lib", "pages", "services", "utils"]);

function isServicePath(filePath: string): boolean {
  return filePath.split("/").some((p) => SERVICE_DIRS.has(p));
}

function emit(
  fullName: string,
  symbolPos: number, symbolEnd: number,
  scopePos: number, scopeEnd: number,
  ctx: DetectorContext,
): Mutation {
  return {
    type: "node",
    kind: "service",
    payload: {
      id: `${ctx.repoKey}::${ctx.filePath}::${fullName}`,
      _full_name: fullName,
      _file: ctx.filePath,
      _symbol_pos: symbolPos,
      _symbol_end: symbolEnd,
      _scope_pos: scopePos,
      _scope_end: scopeEnd,
    },
  };
}

export class ServiceDetector implements Detector {
  name = "service";

  detect(sf: ts.SourceFile, primitives: Primitive[], ctx: DetectorContext): Mutation[] {
    if (!isServicePath(ctx.filePath)) return [];
    if (!ctx.tsMorphSf) return [];
    const tsf = ctx.tsMorphSf;
    const muts: Mutation[] = [];

    for (const decl of tsf.getFunctions()) {
      const name = decl.getName();
      if (!name) continue;
      if (name.startsWith("_")) continue;
      muts.push(emit(
        name,
        decl.getStart(), decl.getEnd(),
        decl.getStart(), decl.getEnd(),
        ctx,
      ));
    }

    for (const stmt of tsf.getVariableStatements()) {
      if (!stmt.isExported()) continue;
      for (const varDecl of stmt.getDeclarations()) {
        const name = varDecl.getName();
        const init = varDecl.getInitializer();
        if (!name || !init) continue;
        if (name.startsWith("_")) continue;
        if (Node.isArrowFunction(init) || Node.isFunctionExpression(init)) {
          muts.push(emit(
            name,
            varDecl.getStart(), varDecl.getEnd(),
            init.getStart(), init.getEnd(),
            ctx,
          ));
        }
      }
    }
    return muts;
  }
}
